crates/compiler/builtins/roc/Json.roc

```diff
@@ -554,91 +506,589 @@ decodeRecord = \initialState, stepField, finalizer -> Decode.custom \bytes, @Jso
             Ok val -> { result: Ok val, rest: afterRecordBytes }
             Err e -> { result: Err e, rest: afterRecordBytes }
 
-decodeTuple = \initialState, stepElem, finalizer -> Decode.custom \initialBytes, @Json {} ->
-        # NB: the stepper function must be passed explicitly until #2894 is resolved.
-        decodeElems = \stepper, state, index, bytes ->
-            { val: newState, rest: beforeCommaOrBreak } <- tryDecode
-                    (
-                        when stepper state index is
-                            TooLong ->
-                                { rest: beforeCommaOrBreak } <- bytes |> anything |> tryDecode
-                                { result: Ok state, rest: beforeCommaOrBreak }
+# JSON NUMBER PRIMITIVE --------------------------------------------------------
+
+# Takes the bytes for a valid Json number primitive into a RocStr
+#
+# Note that this does not handle leading whitespace, any whitespace must be
+# handled in json list or record decoding.
+#
+# |> List.dropIf \b -> b == '+'
+# TODO ^^ not needed if roc supports "1e+2", this supports
+# "+" which is permitted in Json numbers
+#
+# |> List.map \b -> if b == 'E' then 'e' else b
+# TODO ^^ not needed if roc supports "1E2", this supports
+# "E" which is permitted in Json numbers
+takeJsonNumber : List U8 -> { taken : List U8, rest : List U8 }
+takeJsonNumber = \bytes ->
+    when List.walkUntil bytes Start numberHelp is
+        Finish n | Zero n | Integer n | FractionB n | ExponentC n ->
+            taken =
+                bytes
+                |> List.sublist { start: 0, len: n }
+                |> List.dropIf \b -> b == '+'
+                |> List.map \b -> if b == 'E' then 'e' else b
+
+            { taken, rest: List.drop bytes n }
 
-                            Next decoder ->
-                                Decode.decodeWith bytes decoder (@Json {})
-                    )
+        _ ->
+            { taken: [], rest: bytes }
+
+numberHelp : NumberState, U8 -> [Continue NumberState, Break NumberState]
+numberHelp = \state, byte ->
+    when Pair state byte is
```

Comment:
This is so cool! 😻 

Have you tried it with tuples instead of `Pair`?

Reply:
Glad you like it. I was trying to be efficient with state, and make as few (hopefully no) allocations as possible, but still support the various edge cases.

I started with Tuples but hit a few bugs, which have since been fixed by Ayaz. I'll see if I can switch back to tuples as that felt nicer.

Reply:
Converted back to tuples and works a charm. 🎉